tsl/src/partialize_agg.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+
+#include <postgres.h>
+#include <catalog/pg_operator.h>
+#include <miscadmin.h>
+#include <nodes/bitmapset.h>
+#include <nodes/makefuncs.h>
+#include <nodes/nodeFuncs.h>
+#include <optimizer/cost.h>
+#include <optimizer/optimizer.h>
+#include <optimizer/pathnode.h>
+#include <optimizer/paths.h>
+#include <parser/parsetree.h>
+#include <utils/builtins.h>
+#include <utils/lsyscache.h>
+#include <utils/typcache.h>
+#include <utils/fmgroids.h>
+
+#include <planner.h>
+
+#include "nodes/decompress_chunk/decompress_chunk.h"
+#include "partialize_agg.h"
+#include "utils.h"
+#include "debug_assert.h"
+
+#define is_restricted_path(path) list_length(path->parent->baserestrictinfo)
+
+/*
+ * Are we able to optimize the path by applying vectorized aggregation?
+ */
+static bool
+is_vectorizable_agg_path(PlannerInfo *root, AggPath *agg_path, Path *path)
+{
+	Assert(agg_path->aggstrategy == AGG_SORTED || agg_path->aggstrategy == AGG_PLAIN ||
+		   agg_path->aggstrategy == AGG_HASHED);
+
+	/* Having is not supported at the moment */
+	if (root->hasHavingQual)
+		return false;
+
+	/* Only vectorizing within the decompress node is supported so far */
+	bool is_decompress_chunk = ts_is_decompress_chunk_path(path);
+	if (!is_decompress_chunk)
+		return false;
+
+	DecompressChunkPath *decompress_path = (DecompressChunkPath *) path;
+	Assert(decompress_path->custom_path.custom_paths != NIL);
+
+	/*
+	 * It would be great if we could check decompress_path -> have_bulk_decompression_columns
+	 * here (or ensure the column can be bulk decompressed). However, this information is only
+	 * computed when a path is selected and the plan is created.
+	 *
+	 * However, due to the restriction to a particular data type for vectorized aggregates
+	 * (i.e., INT4) we can assume that this column type can be bulk decompressed.
+	 */
+	Path *compressed_path = linitial(decompress_path->custom_path.custom_paths);
+
+	/* No filters are supported at the moment */
+	if (is_restricted_path(path) || is_restricted_path(compressed_path))
+		return false;
```

Comment:
Doesn't this code already work for filters that are entirely pushed down to the compressed path, like a filter on segmentby?

Reply:
I will enable this optimization in a follow-up PR and add proper tests: https://github.com/timescale/timescaledb/pull/6218